src/c2pa/c2pa.py

```diff
@@ -494,203 +730,371 @@ def flush_callback(ctx):
+                    self._own_stream = None
+
+            # Clean up file
+            if hasattr(self, '_file'):
+                try:
+                    self._file.close()
+                except Exception as e:
+                    print(self._error_messages['file_error'].format(str(e)), file=sys.stderr)
+                finally:
+                    self._file = None
+
+            # Clear any stored strings
+            if hasattr(self, '_strings'):
+                self._strings.clear()
+        except Exception as e:
+            print(self._error_messages['cleanup_error'].format(str(e)), file=sys.stderr)
+        finally:
+            self._closed = True
+
     def json(self) -> str:
         """Get the manifest store as a JSON string.
-        
+
         Returns:
             The manifest store as a JSON string
-            
+
         Raises:
             C2paError: If there was an error getting the JSON
         """
+
         if not self._reader:
             raise C2paError("Reader is closed")
         result = _lib.c2pa_reader_json(self._reader)
         return _handle_string_result(result)
-    
+
     def resource_to_stream(self, uri: str, stream: Any) -> int:
         """Write a resource to a stream.
-        
+
         Args:
             uri: The URI of the resource to write
             stream: The stream to write to (any Python stream-like object)
-            
+
         Returns:
             The number of bytes written
-            
+
         Raises:
             C2paError: If there was an error writing the resource
         """
         if not self._reader:
             raise C2paError("Reader is closed")
-        
+
         # Keep uri string alive
         self._uri_str = uri.encode('utf-8')
-        stream_obj = Stream(stream)
-        result = _lib.c2pa_reader_resource_to_stream(self._reader, self._uri_str, stream_obj._stream)
-        
-        if result < 0:
-            _handle_c2pa_error()
-        
-        return result
-    
-    def resource_to_file(self, uri, path) -> None:
-        """Write a resource to a file.
-        """
-        with open(path, "wb") as file:
-            return self.resource_to_stream(uri, file)
+        with Stream(stream) as stream_obj:
+            result = _lib.c2pa_reader_resource_to_stream(self._reader, self._uri_str, stream_obj._stream)
+
+            if result < 0:
+                error = _handle_string_result(_lib.c2pa_error())
+                if error:
+                    raise C2paError(error)
+
+            return result
 
 class Signer:
     """High-level wrapper for C2PA Signer operations."""
-    
+
+    def __init__(self, signer_ptr: ctypes.POINTER(C2paSigner)):
+        """Initialize a new Signer instance.
+
+        Note: This constructor is not meant to be called directly.
+        Use from_info() or from_callback() instead.
+        """
+        self._signer = signer_ptr
+        self._closed = False
+        self._error_messages = {
```

Comment:
We need to huddle on error returns. I know there will be some differences depending on languages, but we also need to have consistency so that we don't have a completely different set of documentation for each binding. The errors are part of the API.

Reply:
Ack - albeit those are specific to the Stream implementation on Python here.